R/requirements_helpers.R

```diff
@@ -11,39 +11,97 @@ legal_r_package_name = function(name){
+  grepl(re_target, existing)
+}
+
+
+compare_compatible_version = function(existing, target) {
+  #' @keywords internal
+  #' Check if existing installed version ~= target version
+  #'
+  #' @param existing The semantic version number installed
+  #' @param target The semantic version number targeted by requirements file
+  #'
+  #' @return TRUE if versions are compatible; otherwise FALSE
+  #'
+  #' @examples
+  #' compare_compatible_version('1.1.1', '1.*')
+  #' # [1] TRUE
+  #' compare_compatible_version('1.1.1', '1.0')
+  #' # [1] TRUE
+  #' compare_compatible_version('1.1.1', '2.0')
+  #' # [1] FALSE
+
+  # Replace trailing version specifier with wildcard (i.e. 2.2 -> 2.*)
+  compatible_version_wildcard = sub('(.*\\.)(.+)', '\\1*', target)
+
+  # Per PEP 440 `~= 2.2` is equivalent to `>= 2.2 & == 2.*`
+  compare_version(existing, compatible_version_wildcard, '==') &
+    compare_version(existing, target, '>=')
 }
 
-check_version = function(target,existing,comp){
-  #' @param target The version targeted for installation
-  #' @param existing The currently installed version
+
+compare_version = function(existing, target, comp) {
+  #' @keywords internal
+  #' Check if existing requirement matches user specified target version
+  #'
+  #' @param existing The semantic version number installed
+  #' @param target The semantic version number targeted by requirements file
   #' @param comp The version comparison operator
-  #' @return logical indicating if existing version matches requirements
-
-  target = strsplit(target,"[^0-9a-zA-Z\\*]")[[1]]
-  existing = strsplit(existing,"[^0-9a-zA-Z\\*]")[[1]]
-  fill = '0'
-  if(target[length(target)] == '*' | comp == '~=') fill = '*'
-  if(length(target) > length(existing)){
-    existing = c(existing,rep(fill,length(target) - length(existing)))
-  } else {
-    target = c(target,rep(fill,length(existing) - length(target)))
-  }
-  to_compare = cbind(existing,target,comp)
-  results = mapply(compare_version,to_compare[,1],to_compare[,2],to_compare[,3])
-  if(comp == '!=') return(any(results)) else return(all(results))
+  #'
+  #' @return logical(1) indicating if comparison is TRUE
+  #'
+  #' @examples
+  #' compare_version('1.0.1', '1.*', '==')
+  #' [1] TRUE
+  #' compare_version('1.0.1', '1.*', '~=')
+  #' [1] TRUE
+  #' compare_version('1.0.1', '1.*', '!=')
+  #' [1] FALSE
+
+  # Push compatible version check to dedicated helper
+  if (comp == "~=") return(compare_compatible_version(existing, target))
+
+  # Check if existing == target would satisfy comparison
+  is_equality_check = comp %in% c('>=', '<=', '~=', '==')
+
+  # Check if `target` is equivalent to `existing` wrt wildcards
+  is_equivalent_version = equivalent_version(existing, target)
+
+  if (is_equality_check & is_equivalent_version) return(TRUE)
+  if (comp == "!=" & is_equivalent_version) return(FALSE)
+
+  # TODO: definitively prove this is a valid substitution
+  non_wild_card_target = gsub('*', '0', target, fixed = TRUE)
```

Comment:
Want to ensure we have all possible edge cases covered in tests.  My brain is too fried rn to be able to think through if this substitution is safe at this point in the function.

Reply:
So if I understand this correctly, the requirement `mgsub <= 2.*` gets converted to `mgsub <= 2.0`, right?

Reply:
The equivalent version check always starts at the front of the string, so you won't have an issue of `existing = '2.9'; target = '2.*'` mucking it up, so it seems like this should be sufficient. Given that all tests pass, it seems solid.